README
## System Overview:

The mongo-connector system is designed to hook up mongoDB to any target system. This allows all the
documents in mongoDB to be stored in some other system, and both mongo and the target system will remain
in sync while the connector is running. It has been tested with python 2.7 and python 3.

## Getting Started:

Since the connector does real time syncing, it is necessary to have MongoDB running, although the
connector will work with both sharded and non sharded configurations. It requires a replica set
setup.

To start the system, simply run "python mongo_connector.py". It is likely, however, that you will need
to specify some command line options to work with your setup. They are described below:

`-m` or `--mongos` is to specify the main address, which is a host:port pair. For sharded clusters,
this should be the mongos address. For individual replica sets, supply the address of the primary.
For example, `-m localhost:27217` would be a valid argument to `-m`. Don't use quotes around the address.

`-t` or `--target-url` is to specify the URL to the target system being used. For example, if you
were using Solr out of the box, you could use '-t http://localhost:8080/solr' with the
SolrDocManager to establish a proper connection. Don't use quotes around address. If target
system doesn't need URL, don't specify.

`-o` or `--oplog-ts` is to specify the name of the file that stores the oplog progress timestamps.
This file is used by the system to store the last timestamp read on a specific oplog. This allows
for quick recovery from failure. By default this is `config.txt`, which starts off empty. An empty
file causes the system to go through all the mongo oplog and sync all the documents. Whenever the
cluster is restarted, it is essential that the oplog-timestamp config file be emptied - otherwise
the connector will miss some documents and behave incorrectly.

`-n` or `--namespace-set` is used to specify the namespaces we want to consider. For example, if we
wished to store all documents from the test.test and alpha.foo namespaces, we could use
`-n test.test,alpha.foo`. The default is to consider all the namespaces, excluding the system and config
databases, and also ignoring the "system.indexes" collection in any database.

`-u` or `--unique-key` is used to specify the mongoDB field that will serve
as the unique key for the target system.  The default is "_id",
which can be noted by "-u _id"

`-f` is to specify a file which contains the password for authentication.
This file is used by mongos to authenticate connections to the shards,
and we'll use it in the oplog threads. The main use of this option is to specify a
password without entering it as plaintext on the command line.

`-p` is to specify the password used for authentication. If this option is specified along
with `-f`, then the password specified here will be used regardless of the contents of the
password file. For sharded clusters, the admin username/password must exist in every shard's 
admin database, otherwise the system will fail to authenticate. This is because of how the 
connector authenticates against the shards, which is described in depth in the System Internals
section below. 

`-a' or `--admin-username` is used to specify the username of an admin user to authenticate with.
To use authentication with the system, the user must specify both this option and the password or password file
option, which stores the password for the user. The default username is '__system', which is not
recommended for production use. If using this and one of the password options in a sharded environment,
it is essential that the username/password exists in the admin database of every shard in the cluster. Otherwise,
authentication will fail. 

`-d` or `--docManager` is used to specify the doc manager file that is going to be used.
You should send the path of the file you want to be used. By default, it will use
the doc_manager_simulator.py file. It is recommended that all doc manager files be kept
in the doc_managers folder in mongo-connector.
For more information about making your own doc manager, see Doc Manager section.

An example of combining all of these is:

	python mongo_connector.py -m localhost:27217 -t http://localhost:8080/solr -o oplog_progress.txt -n alpha.foo,test.test -u _id -k auth.txt -a admin -d ./doc_managers/solr_doc_manager.py

## Usage With Solr:

We have provided an example Solr schema called `schema.xml`, which provides field definitions for the 'name', '_ts', `ns`, and `_id` fields. The schema also sets the `_id` field to be the unique key by adding this line:

     <uniqueKey>_id</uniqueKey>

Solr does not require all the fields to be present, unless a field's `required` attribute is `True`, but the schema must have a line defining every field that may be present in a document field. Some examples of field definitions in the schema are:

    <field name="_id" type="string" indexed="true" stored="true" />
    <field name="_ts" type="long" indexed="true" stored="true" />

The sample XML schema is designed to work with the tests. For a more complete guide to adding fields, review the Solr documentation.

## DocManager

This is the only file that is engine specific. In the current version, we have provided sample
implementations for ElasticSearch, Solr and Mongo, which are in the docmanagers folder.
If you would like to integrate MongoDB with some other engine, then you need to write a
doc manager file for that target system. The sample_doc_manager.py file gives a
detailed description of the functions used by the Doc Manager; the following functions must be
implemented:

__1) init(self)__

This method should if needed, verify the url to the target system and return None if that fails. It should
also create the connection to the target system, and start a periodic
committer if necessary. It can take extra optional parameters for internal use, like
auto_commit.
The unique_key should default to '_id' and it is an obligatory parameter.
It requires a url paramater iff mongo_connector.py is called with the -b paramater.
Otherwise, it doesn't require any other parameter (e.g. if the target engine doesn't need a URL).
It should raise a SystemError exception if the URL is not valid.

__2) stop(self)__

This method must stop any threads running from the DocManager. In some cases this simply stops a
timer thread, whereas in other DocManagers it does nothing because the manager
doesn't use any threads. This method is only called when the MongoConnector is
forced to terminate, either due to errors or as part of normal procedure.

__3) upsert(self, doc)__

Update or insert a document into your engine. The document has ns and _ts fields.
This method should call whatever add/insert/update method exists for
the target system engine and add the document in there. The input will
always be one mongo document, represented as a Python dictionary.
This document will be the current mongo version of the document,
not necessarily the version at the time the upsert was made; the
doc manager will be responsible to track the changes if necessary.
Note this is not necessary to ensure consistency.
It is possible to get two inserts for the same document with the same
contents if there is considerable delay in trailing the oplog.
We have only one function for update and insert because incremental
updates are not supported, so there is no update option.

__4) remove(self, doc)__

Removes document from engine
The input is a python dictionary that represents a mongo document.

__5) search(self, start_ts, end_ts)__

Called to query engine for documents in a time range, including start_ts and end_ts
This method is only used by rollbacks to query all the documents in
the target engine within a certain timestamp window. The input will be two longs
(converted from Bson timestamp) which specify the time range. The 32 most significant
bits are the Unix Epoch Time, and the other bits are the increment. For all purposes,
the function should just do a simple search for timestamps between these values
treating them as simple longs.
The return value should be an iterable set of documents.

__6) commit(self)__

This function is used to force a refresh/commit.
It is used only in the beginning of rollbacks and in test cases, and is
not meant to be called in other circumstances. The body should commit
all documents to the target system (like auto_commit), but not have
any timers or run itself again (unlike auto_commit).  In the event of
too many engine searchers, the commit can be wrapped in a
retry_until_ok to keep trying until the commit goes through.

__7) get_last_doc(self)__

Returns the last document stored in the target engine.
This method is used for rollbacks to establish the rollback window,
which is the gap between the last document on a mongo shard and the
last document. If there are no documents, this functions
returns None. Otherwise, it returns the first document.


## System Internals:

The main Connector thread connects to either a mongod or a mongos, depending on cluster setup, and
spawns an OplogThread for every primary node in the cluster. These OplogThreads continuously poll
the Oplog for new operations, get the relevant documents, and insert and/or update them into the
target system. The general workflow of an OplogThread is:

1. "Prepare for sync", which initializes a tailable cursor to the Oplog of the mongod. This cursor
    will return all new entries, which are processed in step 2. If this is the first time the
    mongo-connector is being run, then a special "dump collection" occurs where all the documents
    from a given namespace are dumped into the target system. On subsequent calls, the timestamp of the
    last oplog entry is stored in the system, so the OplogThread resumes where it left off instead
    of rereading  the whole oplog.

2. For each entry in the oplog, we see if it's an insert/update operation or a delete operation.
   For insert/update operations, we fetch the document from either the mongos or the primary
   (depending on cluster setup). This document is passed up to the DocManager. For deletes, we pass
   up the unique key for the document to the DocManager.

3. The DocManager can either batch documents and periodically update the target system, or immediately,
   or however the user chooses to.

The above three steps essentially loop forever. For usage purposes, the only relevant layer is the
DocManager, which will always get the documents from the underlying layer and is responsible for
adding to/removing from the target system.

Mongo-Connector imports a DocManager from the specified file in mongo_connector.py, preferably
from the doc_managers folder. We have provided sample implementations for a Solr search DocManager
 and an ElasticSearch DocManager. Note that upon execution, mongo_connector.py will copy the file
 to the main folder as doc_manager.py.

The documents stored in the target system are equivalent to what is stored in mongo, except every
document has two additional fields called `ns` and `_ts`. The `_ts` field stores the latest
timestamp corresponding to that document in the oplog. For example, if document `D` was inserted
at time `t1` and then updated at time `t2`, then after processing the update, we update the
documents timestamp to reflect the most recent time. This is relevant for rollbacks in the system.

The `ns` field stores the namespace for each document. This is relevant to prevent conflicts
between identical documents in different namespaces, and also used for managing rollbacks and
syncing.

## Testing scripts

There are two sets of tests. One is for the general system, which can be found in the Tests folder, and the other is for
the Doc Managers, which is found in the tests folder inside the doc_managers folder. The doc manager tests use
the Doc Managers stored in the doc_managers folder; unlike mongo_connector.py they do not copy the file to
doc_manager.py in the main folder.

There are shell scripts for running all tests for each search engine (currenctly Solr, Elastic and Mongo),
for the general connector tests, and a script that runs all tests. You can find them in tests/scripts.

A word of caution: For some of the tests, specifically test_oplog_manager, the output prints out "ERROR:root:OplogThread:
No oplog for thread: Connection('localhost', 27117)". This is expected behavior because the tests run some parts in isolation,
and are not an indication of error. If the tests actually fail, a distinct message will be printed at the very end stating exactly where it failed.

## Troubleshooting

The most common issue when installing is to forget to clear config.txt before restarting the mongo connector.
This config.txt file stores an oplog timestamp which indicates at which point of the oplog the connector should
start. This is made so you can stop syncing at any point by quitting the program and restart from where you left
off later on.

CHANGELOG
Changelog
=========

Version 1.2
-----------

Version 1.2 is a major release with a large number of fixes since the last release on PyPI. It also includes a number of improvements for use with Solr and ElasticSearch.

Improvements
~~~~~~~~~~~~

- Ability to have multiple targets of replication
- Ability to upsert documents containing arrays and nested documents with the Solr DocManager
- Upserts during a collection dump may happen in bulk, resulting in a performance boost
- mongo-connector does not commit writes in target systems by default, resulting in a peformance boost

.. Warning:: This new behavior may give unexpected delays before
             documents are comitted in the target system. Most
             indexing systems provide some way of configuring how
             often changes should be comitted. Please see the relevant
             wiki articles for `Solr
             <https://github.com/10gen-labs/mongo-connector/wiki/Usage%20with%20Solr#managing-commit-behavior/>`_
             and `ElasticSearch
             <https://github.com/10gen-labs/mongo-connector/wiki/Usage%20with%20ElasticSearch#managing-refresh-behavior/>`_
             for more information on configuring commit behavior for
             your system. Note that MongoDB as a target system is
             unaffected by this change.

- Addition of ``auto-commit-interval`` to the command-line options
- Ability to change the destination namespace of upserted documents
- Ability to restrict the fields upserted in documents
- Memory footprint reduced
- Collection dumps may happen in batch, resulting in huge performance gains

Fixes
~~~~~

- Fix for unexpected exit during chunk migrations and orphan documents in MongoDB
- Fix installation problems due to namespace issues

.. Warning:: RENAME of ``mongo_connector.py`` module to
             ``connector.py``. Thus, if you should need to import the
             ``Connector`` object, you now should do
             ``from mongo_connector.connector import Connector``

- Fix user-specified unique keys in Solr and ElasticSearch DocManagers
- Fix for keyboard exit taking large amounts of time to be effective

Version 1.1.1
-------------

This was the first release of mongo-connector.


clustersetup.py
"""
Module with code to setup cluster and test oplog_manager functions.
    
"""

import subprocess
import sys
import time
import os
import json

from pymongo import Connection
from pymongo.errors import ConnectionFailure
from os import path
from mongo_internal import Daemon
from threading import Timer
from oplog_manager import OplogThread
from solr_doc_manager import SolrDocManager
from pysolr import Solr
from util import (long_to_bson_ts, 
                  bson_ts_to_long)
from checkpoint import Checkpoint
from bson.objectid import ObjectId

# Global path variables
PORTS = {"PRIMARY":"27117", "SECONDARY":"27118", "ARBITER":"27119", 
    "CONFIG":"27220", "MONGOS":"27217"}
SETUP_DIR = path.expanduser("~/mongo-connector/test")
DEMO_SERVER_DATA = SETUP_DIR + "/data"
DEMO_SERVER_LOG = SETUP_DIR + "/logs"
MONGOD_KSTR = " --dbpath " + DEMO_SERVER_DATA
MONGOS_KSTR = "mongos --port " + PORTS["CONFIG"]


def remove_dir(path):
    """Remove supplied directory"""
    command = ["rm", "-rf", path]
    subprocess.Popen(command).communicate()

def create_dir(path):
    """Create supplied directory"""
    command = ["mkdir", "-p", path]
    subprocess.Popen(command).communicate()


def killMongoProc(host, port):
    try:
        conn = Connection(host, int(port))
        conn['admin'].command('shutdown',1, force=True)
    except:
        cmd = ["pgrep -f \"" + str(port) + MONGOD_KSTR + "\" | xargs kill -9"]
        executeCommand(cmd)

def killMongosProc():
    cmd = ["pgrep -f \"" + MONGOS_KSTR + "\" | xargs kill -9"]
    executeCommand(cmd)

def killAllMongoProc(host):
    """Kill any existing mongods"""
    for port in PORTS.values():
        killMongoProc(host, port)

def startMongoProc(port, replSetName, data, log):
    #Create the replica set
    CMD = ["mongod --fork --replSet " + replSetName + " --noprealloc --port " + port + " --dbpath "
    + DEMO_SERVER_DATA + data + " --shardsvr --rest --logpath "
    + DEMO_SERVER_LOG + log + " --logappend &"]
    
    executeCommand(CMD)
    checkStarted(int(port))


def executeCommand(command):
    """Wait a little and then execute shell command"""
    time.sleep(1)
    #return os.system(command)
    subprocess.Popen(command, shell=True)


#========================================= #
#   Helper functions to make sure we move  #
#   on only when we're good and ready      #
#========================================= #


def tryConnection(port):
    """Uses pymongo to try to connect to mongod"""
    error = 0
    try:
        Connection('localhost', port)
    except Exception:
        error = 1
    return error


def checkStarted(port):
    """Checks if our the mongod has started"""
    connected = False

    while not connected:
        error = tryConnection(port)
        if error:
            #Check every 1 second
            time.sleep(1)
        else:
            connected = True


#================================= #
#       Run Mongo* processes       #
#================================= #

    

class ReplSetManager():

    def startCluster(self):
        # Kill all spawned mongods
        killAllMongoProc('localhost')

        # Kill all spawned mongos
        killMongosProc()
        
        
        remove_dir(DEMO_SERVER_LOG)
        remove_dir(DEMO_SERVER_DATA)        

        create_dir(DEMO_SERVER_DATA + "/standalone/journal")
        create_dir(DEMO_SERVER_DATA + "/replset1a/journal")
        create_dir(DEMO_SERVER_DATA + "/replset1b/journal")
        create_dir(DEMO_SERVER_DATA + "/replset1c/journal")
        create_dir(DEMO_SERVER_DATA + "/shard1a/journal")
        create_dir(DEMO_SERVER_DATA + "/shard1b/journal")
        create_dir(DEMO_SERVER_DATA + "/config1/journal")
        create_dir(DEMO_SERVER_LOG)
            

        # Create the replica set
        startMongoProc(PORTS["PRIMARY"], "demo-repl", "/replset1a", "/replset1a.log")
        startMongoProc(PORTS["SECONDARY"], "demo-repl", "/replset1b", "/replset1b.log")
        startMongoProc(PORTS["ARBITER"], "demo-repl", "/replset1c", "/replset1c.log")
        
        # Setup config server
        CMD = ["mongod --oplogSize 500 --fork --configsvr --noprealloc --port " + PORTS["CONFIG"] + " --dbpath " +
        DEMO_SERVER_DATA + "/config1 --rest --logpath "
       + DEMO_SERVER_LOG + "/config1.log --logappend &"]
        executeCommand(CMD)
        checkStarted(int(PORTS["CONFIG"]))

# Setup the mongos
        CMD = ["mongos --port " + PORTS["MONGOS"] + " --fork --configdb localhost:" + PORTS["CONFIG"] + " --chunkSize 1  --logpath " 
       + DEMO_SERVER_LOG + "/mongos1.log --logappend &"]
        executeCommand(CMD)    
        checkStarted(int(PORTS["MONGOS"]))
        
# Configure the shards and begin load simulation
        #executeCommand(CMD).wait()
        cmd1 = "mongo --port " + PORTS["PRIMARY"] + " " + SETUP_DIR + "/setup/configReplSet.js"
        cmd2 = "mongo --port  "+ PORTS["MONGOS"] + " " + SETUP_DIR + "/setup/configMongos.js"
        
        subprocess.call(cmd1, shell=True)
        conn = Connection('localhost:' + PORTS["PRIMARY"])
        sec_conn = Connection('localhost:' + PORTS["SECONDARY"])
        
        while conn['admin'].command("isMaster")['ismaster'] is False:
            time.sleep(1)
        
        while sec_conn['admin'].command("replSetGetStatus")['myState'] != 2:
            time.sleep(1)
        
        subprocess.call(cmd2, shell=True)
    
    def abort_test(self):
        print 'test failed'
        sys.exit(1)
    
    def test_mongo_internal(self):
        t = Timer(60, self.abort_test)
        t.start()
        d = Daemon('localhost:' + PORTS["MONGOS"], None)
        d.start()
        while len(d.shard_set) == 0:
            pass
        t.cancel()
                    
        d.stop()
        #the Daemon should recognize a single running shard
        assert len(d.shard_set) == 1
        #we want to add several shards
        
    def get_oplog_thread(self):
        #try sending invalid entry
        
        
        primary_conn = Connection('localhost', int(PORTS["PRIMARY"]))
        if primary_conn['admin'].command("isMaster")['ismaster'] is False:
            primary_conn = Connection('localhost', int(PORTS["SECONDARY"]))
        
        primary_conn['test']['test'].drop()
        mongos_conn = "localhost:" + PORTS["MONGOS"]
        
        oplog_coll = primary_conn['local']['oplog.rs']
        oplog_coll.drop()           # reset the oplog
        
        primary_conn['local'].create_collection('oplog.rs', capped=True, size=1000000)
        namespace_set = ['test.test']
        doc_manager = SolrDocManager('http://localhost:8080/solr', False)
        oplog = OplogThread(primary_conn, mongos_conn, oplog_coll, True, doc_manager, None, namespace_set)
        
        return (oplog, primary_conn, oplog_coll)

    def get_oplog_thread_new(self):
        #try sending invalid entry
        
        primary_conn = Connection('localhost', int(PORTS["PRIMARY"]))
        if primary_conn['admin'].command("isMaster")['ismaster'] is False:
            primary_conn = Connection('localhost', int(PORTS["SECONDARY"]))

        mongos_conn = "localhost:" + PORTS["MONGOS"]
        oplog_coll = primary_conn['local']['oplog.rs']
        
        namespace_set = ['test.test']
        doc_manager = SolrDocManager('http://localhost:8080/solr', False)
        oplog = OplogThread(primary_conn, mongos_conn, oplog_coll, True, doc_manager, None, namespace_set)
        
        return (oplog, primary_conn, oplog_coll)

            
    def test_retrieve_doc(self):
        
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread()
        #testing for entry as none type
        entry = None
        assert (test_oplog.retrieve_doc(entry) is None)
        
        oplog_cursor = oplog_coll.find({},tailable=True, await_data=True)
        #oplog_cursor.next()  #skip first 'drop collection' operation
        
        
        primary_conn['test']['test'].insert ( {'name':'paulie'})
        last_oplog_entry = oplog_cursor.next()
        target_entry = primary_conn['test']['test'].find_one()
        
        #testing for search after inserting a document
        assert (test_oplog.retrieve_doc(last_oplog_entry) == target_entry)
        
        primary_conn['test']['test'].update({'name':'paulie'}, {"$set": {'name':'paul'}} )
        
        last_oplog_entry = oplog_cursor.next()
        target_entry = primary_conn['test']['test'].find_one()        
        
        #testing for search after updating a document
        assert (test_oplog.retrieve_doc(last_oplog_entry) == target_entry)
        
        primary_conn['test']['test'].remove( {'name':'paul'} )
        last_oplog_entry = oplog_cursor.next()
        
        #testing for search after deleting a document
        assert (test_oplog.retrieve_doc(last_oplog_entry) == None)
        
        last_oplog_entry['o']['_id'] = 'badID'
        
        #testing for bad doc id as input
        assert (test_oplog.retrieve_doc(last_oplog_entry) == None)
        
        test_oplog.stop()
        
                
    def test_get_oplog_cursor(self):
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread()
        
        #test None cursor
        assert (test_oplog.get_oplog_cursor(None) == None)
        
        #test with one document
        primary_conn['test']['test'].insert ( {'name':'paulie'} )
        ts = test_oplog.get_last_oplog_timestamp()
        cursor = test_oplog.get_oplog_cursor(ts)
        assert (cursor.count() == 1)
        
        #test with two documents, one after the ts
        primary_conn['test']['test'].insert ( {'name':'paul'} )
        cursor = test_oplog.get_oplog_cursor(ts)
        assert (cursor.count() == 2)
        
        #test case where timestamp is not in oplog which implies we're too behind 
        oplog_coll = primary_conn['local']['oplog.rs']
        oplog_coll.drop()           # reset the oplog
        primary_conn['local'].create_collection('oplog.rs', capped=True, size=10000)
        
        primary_conn['test']['test'].insert ( {'name':'pauline'} )
        assert (test_oplog.get_oplog_cursor(ts) == None)
        test_oplog.stop()
            
        #need to add tests for 'except' part of get_oplog_cursor
            

    def test_get_last_oplog_timestamp(self):
        
        #test empty oplog
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread()
        assert (test_oplog.get_last_oplog_timestamp() == None)
        
        #test non-empty oplog
        oplog_cursor = oplog_coll.find({},tailable=True, await_data=True)
        primary_conn['test']['test'].insert ( {'name':'paulie'} )
        last_oplog_entry = oplog_cursor.next()
        assert (test_oplog.get_last_oplog_timestamp() == last_oplog_entry['ts'])
        
        test_oplog.stop()
        
            
    def test_dump_collection(self):
        
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread()
        solr_url = "http://localhost:8080/solr"
        solr = Solr(solr_url)
        solr.delete(q='*:*')
        
        #for empty oplog, no documents added
        assert (test_oplog.dump_collection(None) == None)
        assert (len(solr.search('*')) == 0)
        
        #with documents
        primary_conn['test']['test'].insert ( {'name':'paulie'} )
        search_ts = test_oplog.get_last_oplog_timestamp()
        test_oplog.dump_collection(search_ts)

        test_oplog.doc_manager.commit()
        solr_results = solr.search('*')
        assert (len(solr_results) == 1)
        solr_doc = solr_results.docs[0]
        assert (long_to_bson_ts(solr_doc['_ts']) == search_ts)
        assert (solr_doc['name'] == 'paulie')
        assert (solr_doc['ns'] == 'test.test')
        
        test_oplog.stop()
        
    def test_init_cursor(self):
        
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread()
        test_oplog.checkpoint = Checkpoint()            #needed for these tests
        
        #initial tests with no config file and empty oplog
        assert (test_oplog.init_cursor() == None)
        
        #no config, single oplog entry
        primary_conn['test']['test'].insert ( {'name':'paulie'} )
        search_ts = test_oplog.get_last_oplog_timestamp()
        cursor = test_oplog.init_cursor()
        
        
        assert (cursor.count() == 1)
        assert (test_oplog.checkpoint.commit_ts == search_ts)
        
        #with config file, assert that size != 0
        os.system('touch temp_config.txt')
        
        config_file_path = os.getcwd() + '/'+ 'temp_config.txt'
        test_oplog.oplog_file = config_file_path
        cursor = test_oplog.init_cursor()
        config_file_size = os.stat(config_file_path)[6]
        
        assert (cursor.count() == 1)
        assert (config_file_size != 0)
        
        os.system('rm temp_config.txt')
        test_oplog.stop()
        #testing for valid entry
    
    def test_prepare_for_sync(self):
        
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread()
        cursor = test_oplog.prepare_for_sync()
        
        assert (test_oplog.checkpoint != None)
        assert (test_oplog.checkpoint.commit_ts == None)
        assert (cursor == None)
        
        primary_conn['test']['test'].insert ( {'name':'paulie'} )
        cursor = test_oplog.prepare_for_sync()
        search_ts = test_oplog.get_last_oplog_timestamp()

        #make sure that the cursor is valid and the timestamp is updated properly
        assert (test_oplog.checkpoint.commit_ts == search_ts)
        assert (cursor != None)
        assert (cursor.count() == 1)
        
        primary_conn['test']['test'].insert ( {'name':'paulter'} )
        cursor = test_oplog.prepare_for_sync()
        new_search_ts = test_oplog.get_last_oplog_timestamp()
        
        #make sure that the newest document is in the cursor.
        assert (cursor.count() == 2)
        next_doc = cursor.next()
        assert (next_doc['o']['name'] == 'paulter')
        assert (next_doc['ts'] == new_search_ts)

        test_oplog.stop()
        
    def test_write_config(self):
        
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread()
        test_oplog.checkpoint = Checkpoint()
        
        
        #test that None is returned if there is no config file specified.
        assert (test_oplog.write_config() == None)
        
        #create config file
        os.system('touch temp_config.txt')
        config_file_path = os.getcwd() + '/temp_config.txt'
        test_oplog.oplog_file = config_file_path
        
        primary_conn['test']['test'].insert ( {'name':'paulie'} )
        search_ts = test_oplog.get_last_oplog_timestamp()
        test_oplog.checkpoint.commit_ts = search_ts
        
        test_oplog.write_config()
        
        data = json.load(open(config_file_path, 'r'))
        oplog_str = str(oplog_coll.database.connection)
        
        assert (oplog_str in data[0])
        assert (search_ts == long_to_bson_ts(data[1]))
        
        #ensure that the temporary file is deleted
        assert (os.path.exists(config_file_path + '~') is False)
        
        search_ts = long_to_bson_ts(111111111111111)
        test_oplog.checkpoint.commit_ts = search_ts
        
        test_oplog.write_config()
        data = json.load(open(config_file_path, 'r'))
        oplog_str = str(oplog_coll.database.connection)
        
        assert (oplog_str in data[0])
        assert (search_ts == long_to_bson_ts(data[1]))
        
        test_oplog.stop()
        os.system('rm ' + config_file_path)
        
    def test_read_config(self):
        
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread()
        test_oplog.checkpoint = Checkpoint()
        
        #testing with no file
        assert(test_oplog.read_config() is None)
        #create config file
        os.system('touch temp_config.txt')
        config_file_path = os.getcwd() + '/temp_config.txt'
        test_oplog.oplog_file = config_file_path
        
        #testing with empty file
        assert(test_oplog.read_config() is None)
        
        #testing with a non-empty file
        search_ts = long_to_bson_ts(111111111111111)
        test_oplog.checkpoint.commit_ts = search_ts
        test_oplog.write_config()
        assert (test_oplog.read_config() == search_ts)

        #testing update
        search_ts = long_to_bson_ts(999999999999999)
        test_oplog.checkpoint.commit_ts = search_ts
        test_oplog.write_config()
        assert (test_oplog.read_config() == search_ts)
        os.system('rm ' + config_file_path)   
        
    def test_rollback(self):
        
        test_oplog, primary_conn, oplog_coll = self.get_oplog_thread_new()
        #test_oplog.start()
        
        solr = Solr('http://localhost:8080/solr')
        solr.delete(q='*:*')
        obj1 = ObjectId('4ff74db3f646462b38000001')
        primary_conn['test']['test'].remove({})
        primary_conn['test']['test'].insert( {'_id':obj1,'name':'paulie'},w=2 )
        cutoff_ts = test_oplog.get_last_oplog_timestamp()
       
        obj2 = ObjectId('4ff74db3f646462b38000002')
        first_doc = {'name':'paulie', '_ts':bson_ts_to_long(cutoff_ts), 'ns':'test.test', '_id': obj1}
        
        primary_port = primary_conn.port
        char = None
        altchar = None
        
        if primary_port == int(PORTS["PRIMARY"]):
            secondary_port = int(PORTS["SECONDARY"])
            altchar = "b"
            char = "a"
        else:
            secondary_port = int(PORTS["PRIMARY"])
            altchar = "a"
            char = "b"
        #try kill one, try restarting
                
        
        killMongoProc(primary_conn.host, secondary_port)
                
        primary_conn['test']['test'].insert ( {'_id': obj2, 'name':'paul'})

        killMongoProc(primary_conn.host, primary_port)  
       	
        startMongoProc(str(secondary_port), "demo-repl", "/replset1" + altchar, "/replset1" + altchar + ".log")
        new_primary_conn = Connection(primary_conn.host, secondary_port)
        while new_primary_conn['admin'].command("isMaster")['ismaster'] is False:
            time.sleep(1)
        startMongoProc(str(primary_port), "demo-repl", "/replset1" + char, "/replset1" + char + ".log")
        new_secondary_conn = Connection(primary_conn.host, primary_port)
        while new_secondary_conn['admin'].command("replSetGetStatus")['myState'] != 2:
            time.sleep(1)
        
        assert (new_primary_conn.port == secondary_port)
        assert (new_secondary_conn.port == primary_port)
       
        last_ts = test_oplog.get_last_oplog_timestamp()    
        second_doc = {'name':'paul', '_ts':bson_ts_to_long(last_ts), 'ns':'test.test', '_id': obj2}
        
        test_oplog.doc_manager.upsert([first_doc, second_doc])
        test_oplog.doc_manager.commit()
        test_oplog.rollback()
        test_oplog.doc_manager.commit()
        results = solr.search('*')
        
        assert (len(results) == 1)

        results_doc = results.docs[0]
        assert(results_doc['name'] == 'paulie')
        assert(results_doc['_ts'] == bson_ts_to_long(cutoff_ts))


#test_oplog.stop()
                #        solr.delete(q='*:*')                #empty solr 
                
        """
            { "ts" : { "t" : 1341343596000, "i" : 1 }, "h" : NumberLong(0), "op" : "n", "ns" : "", "o" : { "msg" : "initiating set" } }
            { "ts" : { "t" : 1341347358000, "i" : 1 }, "h" : NumberLong("5704011229614309393"), "op" : "i", "ns" : "test.test", "o" : { "_id" : ObjectId("4ff3561efb3d8f91f511da5c"), "name" : "paulie" } }
            { "ts" : { "t" : 1341347475000, "i" : 1 }, "h" : NumberLong("-1269638662509784076"), "op" : "u", "ns" : "test.test", "o2" : { "_id" : ObjectId("4ff3561efb3d8f91f511da5c") }, "o" : { "$set" : { "name" : "paul" } } }
            { "ts" : { "t" : 1341347486000, "i" : 1 }, "h" : NumberLong("570562443735405485"), "op" : "d", "ns" : "test.test", "b" : true, "o" : { "_id" : ObjectId("4ff3561efb3d8f91f511da5c") } }
        """
        

"""
# Create the sharded cluster
CMD = ["mongod --oplogSize 500 --fork --shardsvr --noprealloc --port 27218 "
       "--dbpath "
       + DEMO_SERVER_DATA + "/shard1a --rest --logpath "
       + DEMO_SERVER_LOG + "/shard1a.log --logappend &"]
executeCommand(CMD)
checkStarted(27218)


CMD = ,["mongod --oplogSize 500 --fork --shardsvr --noprealloc --port 27219 "
       "--dbpath "
       + DEMO_SERVER_DATA + "/shard1b --rest --logpath "
       + DEMO_SERVER_LOG + "/shard1b.log --logappend &"]
executeCommand(CMD)
checkStarted(27219)

"""
